src/graph/graphcontext.c

```diff
@@ -250,35 +252,50 @@ uint GraphContext_AttributeCount(GraphContext *gc) {
 }
 
 Attribute_ID GraphContext_FindOrAddAttribute(GraphContext *gc, const char *attribute) {
+	// Acquire a read lock for looking up the attribute.
+	pthread_rwlock_rdlock(&gc->_attribute_rwlock);
+
 	// See if attribute already exists.
-	Attribute_ID *pAttribute_id = NULL;
-	Attribute_ID attribute_id = GraphContext_GetAttributeID(gc, attribute);
-
-	if(attribute_id == ATTRIBUTE_NOTFOUND) {
-		attribute_id = raxSize(gc->attributes);
-		pAttribute_id = rm_malloc(sizeof(Attribute_ID));
-		*pAttribute_id = attribute_id;
-
-		raxInsert(gc->attributes,
-				  (unsigned char *)attribute,
-				  strlen(attribute),
-				  pAttribute_id,
-				  NULL);
-		gc->string_mapping = array_append(gc->string_mapping, rm_strdup(attribute));
+	void *attribute_id = raxFind(gc->attributes, (unsigned char *)attribute, strlen(attribute));
+
+	if(attribute_id == raxNotFound) {
+		// We are writing to the shared GraphContext; release the held lock and re-acquire as a writer.
+		pthread_rwlock_unlock(&gc->_attribute_rwlock);
+		pthread_rwlock_wrlock(&gc->_attribute_rwlock);
+
+		attribute_id = (void *)raxSize(gc->attributes);
```

Comment:
This is problematic, consider 2 threads all trying to introduce the same set of attributes: ('a', 'b', 'c')
if thread 1 managed to acquire the lock and insert all 3 attributes, while thread 2 is trying to insert attribute 'a'  once 2 gets the lock (trying to introduce 'a') rax size is 3 but the attribute id for 'a' is 1.